staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -766,49 +807,79 @@ func (r *Request) Stream(ctx context.Context) (io.ReadCloser, error) {
 		return nil, err
 	}
 
-	url := r.URL().String()
-	req, err := http.NewRequest(r.verb, url, nil)
-	if err != nil {
-		return nil, err
-	}
-	if r.body != nil {
-		req.Body = ioutil.NopCloser(r.body)
-	}
-	req = req.WithContext(ctx)
-	req.Header = r.headers
 	client := r.c.Client
 	if client == nil {
 		client = http.DefaultClient
 	}
-	r.backoff.Sleep(r.backoff.CalculateBackoff(r.URL()))
-	resp, err := client.Do(req)
-	updateURLMetrics(ctx, r, resp, err)
-	if r.c.base != nil {
+
+	var retryAfter *RetryAfter
+	url := r.URL().String()
+	for {
+		req, err := r.newHTTPRequest(ctx)
 		if err != nil {
-			r.backoff.UpdateBackoff(r.URL(), err, 0)
-		} else {
-			r.backoff.UpdateBackoff(r.URL(), err, resp.StatusCode)
+			return nil, err
+		}
+		if r.body != nil {
+			req.Body = ioutil.NopCloser(r.body)
 		}
-	}
-	if err != nil {
-		return nil, err
-	}
 
-	switch {
-	case (resp.StatusCode >= 200) && (resp.StatusCode < 300):
-		handleWarnings(resp.Header, r.warningHandler)
-		return resp.Body, nil
+		r.backoff.Sleep(r.backoff.CalculateBackoff(r.URL()))
+		if retryAfter != nil {
+			// We are retrying the request that we already send to apiserver
+			// at least once before.
+			// This request should also be throttled with the client-internal rate limiter.
+			if err := r.tryThrottleWithInfo(ctx, retryAfter.Reason); err != nil {
+				return nil, err
+			}
+			retryAfter = nil
+		}
 
-	default:
-		// ensure we close the body before returning the error
-		defer resp.Body.Close()
+		resp, err := client.Do(req)
+		updateURLMetrics(ctx, r, resp, err)
+		if r.c.base != nil {
+			if err != nil {
+				r.backoff.UpdateBackoff(r.URL(), err, 0)
+			} else {
+				r.backoff.UpdateBackoff(r.URL(), err, resp.StatusCode)
+			}
+		}
+		if err != nil {
+			// we only retry on an HTTP response with 'Retry-After' header
+			return nil, err
+		}
+
+		switch {
+		case (resp.StatusCode >= 200) && (resp.StatusCode < 300):
+			handleWarnings(resp.Header, r.warningHandler)
+			return resp.Body, nil
+
+		default:
+			done, err := func() (bool, error) {
+				defer resp.Body.Close()
+
+				var retry bool
+				retryAfter, retry = r.retry.NextRetry(req, resp, err, func(_ *http.Request, _ error) bool {
```

Comment:
So stream would now retry on 5xx errors on non-GET requests?  I'm concerned about side-effects of partially complete requests.

I could probably buy 503 given its usage for, "you didn't get to your endpoint" from the aggregator.

Reply:
it would retry only if `StatusCode == 429 or StatusCode >= 500`  and the response has a `Retry-After` header with an integer value. 
If we see  a `Retry-After` response header, it must have been deliberately put there by the server, so it should be safe to retry. (This is in keeping with the current retry behavior of `Do`)

Reply:
The existence of "Retry-After" header is important.

Kubernetes apiserver sets it only on 429 IIRC, so do we even have a chance to retry on 500+ ?

Reply:
I think we resolved this in our slack conversation.

Reply:
Aggre.